presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/ExpressionRewriteRuleSet.java

```diff
@@ -0,0 +1,269 @@
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableSet;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+
+import static com.facebook.presto.sql.planner.iterative.PlanNodePatterns.aggregation;
+import static com.facebook.presto.sql.planner.iterative.PlanNodePatterns.filter;
+import static com.facebook.presto.sql.planner.iterative.PlanNodePatterns.join;
+import static com.facebook.presto.sql.planner.iterative.PlanNodePatterns.project;
+import static com.facebook.presto.sql.planner.iterative.PlanNodePatterns.tablesScan;
+import static com.facebook.presto.sql.planner.iterative.PlanNodePatterns.values;
+
+public class ExpressionRewriteRuleSet
+        implements RuleSet
+{
+    interface ExpressionRewriteRule
+    {
+        Expression rewrite(Expression expression, Rule.Context context);
+    }
+
+    private final ExpressionRewriteRule rewriter;
+
+    protected ExpressionRewriteRuleSet(ExpressionRewriteRule rewrite)
+    {
+        this.rewriter = rewrite;
+    }
+
+    public Set<Rule> rules()
+    {
+        return ImmutableSet.of(
+                new ProjectExpressionRewrite(),
+                new AggregationExpressionRewrite(),
+                new FilterExpressionRewrite(),
+                new TableScanExpressionRewrite(),
+                new JoinExpressionRewrite(),
+                new ValuesExpressionRewrite(),
+                new ApplyExpressionRewrite());
+    }
+
+    private final class ProjectExpressionRewrite
+            implements Rule
+    {
+        @Override
+        public Pattern getPattern()
+        {
+            return project();
+        }
+
+        @Override
+        public Optional<PlanNode> apply(PlanNode node, Context context)
+        {
+            ProjectNode projectNode = (ProjectNode) node;
+            Assignments assignments = projectNode.getAssignments().rewrite(x -> ExpressionRewriteRuleSet.this.rewriter.rewrite(x, context));
+            if (projectNode.getAssignments().equals(assignments)) {
+                return Optional.empty();
+            }
+            return Optional.of(new ProjectNode(node.getId(), projectNode.getSource(), assignments));
+        }
+    }
+
+    private final class AggregationExpressionRewrite
+            implements Rule
+    {
+        @Override
+        public Pattern getPattern()
+        {
+            return aggregation();
+        }
+
+        @Override
+        public Optional<PlanNode> apply(PlanNode node, Context context)
+        {
+            AggregationNode aggregationNode = (AggregationNode) node;
+            boolean anyRewritten = false;
+            ImmutableMap.Builder<Symbol, Aggregation> aggregations = ImmutableMap.builder();
+            for (Map.Entry<Symbol, Aggregation> aggregation : aggregationNode.getAggregations().entrySet()) {
+                FunctionCall call = (FunctionCall) rewriter.rewrite(aggregation.getValue().getCall(), context);
```

Comment:
If the aggregation function arguments are constrained to be symbol references, maybe the best thing would be to not rewrite it here at all?